File: Packages/js-common/Source/Commands/AddMap.ts
import {E} from "js-vextensions";
import {AssertV, AssertValidate, Command, CommandMeta, DBHelper, dbp, GenerateUUID, SimpleSchema, UUID} from "mobx-graphlink";
import {MapEdit} from "../CommandMacros/MapEdit.js";
import {UserEdit} from "../CommandMacros/UserEdit.js";
import {DMap} from "../DB/maps/@Map.js";
import {NodePhrasing} from "../DB/nodePhrasings/@NodePhrasing.js";
import {NodeL1} from "../DB/nodes/@Node.js";
import {NodeRevision} from "../DB/nodes/@NodeRevision.js";
import {NodeType} from "../DB/nodes/@NodeType.js";
import {GetUserHidden} from "../DB/userHiddens.js";
import {AddChildNode} from "./AddChildNode.js";
import {AddNode} from "./AddNode.js";
import {GetFinalAccessPolicyForNewEntry} from "../DB.js";

@UserEdit
@CommandMeta({
	payloadSchema: ()=>SimpleSchema({
		$map: {$ref: "DMap"},
	}),
	returnSchema: ()=>SimpleSchema({$id: {$ref: "UUID"}}),
})
export class AddMap extends Command<{map: DMap}, {id: UUID}> {
	sub_addNode: AddNode;
	Validate() {
		const {map} = this.payload;
		AssertV(map.featured === undefined, 'Cannot set "featured" to true while first adding a map. (hmmm)');

		map.id = this.GenerateUUID_Once("id");
		map.creator = this.userInfo.id;
		map.createdAt = Date.now();
		map.edits = 0;
		map.editedAt = map.createdAt;

		const userHidden = GetUserHidden.NN(this.userInfo.id);
		//const accessPolicy_rootNode = GetFinalAccessPolicyForNewEntry(null, map.nodeAccessPolicy ?? map.accessPolicy, "nodes");
		const accessPolicy_rootNode = GetFinalAccessPolicyForNewEntry(null, map.accessPolicy, "nodes"); // add-map dialog doesn't let user choose node-access-policy yet, so use the map's accessor policy for the root-node
		const newRootNode = new NodeL1({
			accessPolicy: accessPolicy_rootNode.id,
			type: NodeType.category, creator: map.creator, rootNodeForMap: map.id,
		});
		//const newRootNodeRevision = new NodeRevision(E(map.nodeDefaults, {phrasing: NodePhrasing.Embedded({text_base: "Root"})}));
		const newRootNodeRevision = new NodeRevision({phrasing: NodePhrasing.Embedded({text_base: "Root"})});
		this.IntegrateSubcommand(()=>this.sub_addNode, null, ()=>new AddNode({mapID: map.id, node: newRootNode, revision: newRootNodeRevision}));

		map.rootNode = this.sub_addNode.payload.node.id;
		AssertValidate("DMap", map, "Map invalid");

		this.returnData = {id: map.id};
	}

	DeclareDBUpdates(db: DBHelper) {
		const {map} = this.payload;
		db.set(dbp`maps/${map.id}`, map);
		db.add(this.sub_addNode.GetDBUpdates(db)); // add node first, since map has fk-ref to it
	}
}

File: Packages/app-server/src/db/commands/add_map.rs
use std::fmt::{Display, Formatter};

use rust_shared::anyhow::{anyhow, Context, Error};
use rust_shared::async_graphql::Object;
use rust_shared::async_graphql::{InputObject, SimpleObject, ID};
use rust_shared::db_constants::SYSTEM_USER_ID;
use rust_shared::rust_macros::wrap_slow_macros;
use rust_shared::serde::{Deserialize, Serialize};
use rust_shared::serde_json::{json, Value};
use rust_shared::utils::general_::extensions::ToOwnedV;
use rust_shared::utils::time::time_since_epoch_ms_i64;
use rust_shared::utils::type_aliases::JSONValue;
use rust_shared::{anyhow, async_graphql, serde_json, GQLError};
use tracing::info;

use crate::db::commands::_command::command_boilerplate;
use crate::db::commands::_shared::increment_edit_counts::increment_edit_counts_if_valid;
use crate::db::general::sign_in_::jwt_utils::{get_user_info_from_gql_ctx, resolve_jwt_to_user_info};
use crate::db::map_node_edits::{ChangeType, MapNodeEdit};
use crate::db::maps::{Map, MapInput};
use crate::db::node_phrasings::NodePhrasing_Embedded;
use crate::db::node_revisions::{NodeRevision, NodeRevisionInput};
use crate::db::nodes_::_node::NodeInput;
use crate::db::nodes_::_node_type::NodeType;
use crate::db::users::User;
use crate::utils::db::accessors::AccessorContext;
use crate::utils::general::data_anchor::DataAnchorFor1;
use rust_shared::utils::db::uuid::new_uuid_v4_as_b64;

use super::_command::{insert_db_entry_by_id_for_struct, tbd, upsert_db_entry_by_id_for_struct, CanOmit, NoExtras};
use super::_shared::add_node::add_node;

wrap_slow_macros! {

#[derive(Default)] pub struct MutationShard_AddMap;
#[Object] impl MutationShard_AddMap {
	async fn add_map(&self, gql_ctx: &async_graphql::Context<'_>, input: AddMapInput, only_validate: Option<bool>) -> Result<AddMapResult, GQLError> {
		command_boilerplate!(gql_ctx, input, only_validate, add_map);
	}
}

#[derive(InputObject, Serialize, Deserialize)]
pub struct AddMapInput {
	pub map: MapInput,
}

#[derive(SimpleObject, Debug)]
pub struct AddMapResult {
	pub id: String,
	pub root_node_id: String,
}

}

pub async fn add_map(ctx: &AccessorContext<'_>, actor: &User, _is_root: bool, input: AddMapInput, _extras: NoExtras) -> Result<AddMapResult, Error> {
	let AddMapInput { map: map_ } = input;

	let root_node_id = new_uuid_v4_as_b64();
	let map = Map {
		// set by server
		id: ID(new_uuid_v4_as_b64()),
		creator: actor.id.to_string(),
		createdAt: time_since_epoch_ms_i64(),
		edits: 0,
		editedAt: Some(time_since_epoch_ms_i64()), // field should probably be non-nullable, since we set it at start to begin with
		rootNode: root_node_id.clone(),
		// pass-through
		accessPolicy: map_.accessPolicy,
		name: map_.name,
		note: map_.note,
		noteInline: map_.noteInline,
		defaultExpandDepth: map_.defaultExpandDepth,
		nodeAccessPolicy: map_.nodeAccessPolicy,
		featured: map_.featured,
		editors: map_.editors,
		extras: map_.extras,
	};
	insert_db_entry_by_id_for_struct(&ctx, "maps".o(), map.id.to_string(), map.clone()).await?;

	let new_root_node = NodeInput {
		accessPolicy: map.accessPolicy.clone(), // add-map dialog doesn't let user choose node-access-policy yet, so use the map's accessor policy for the root-node
		r#type: NodeType::category,
		rootNodeForMap: Some(map.id.to_string()),
		multiPremiseArgument: None,
		argumentType: None,
		extras: CanOmit::None,
	};
	let new_root_node_rev = NodeRevisionInput {
		phrasing: NodePhrasing_Embedded { text_base: "Root".o(), text_negation: None, text_question: None, text_narrative: None, note: None, terms: vec![] },
		node: None,
		displayDetails: None,
		attachments: vec![],
	};
	add_node(ctx, actor, new_root_node, Some(root_node_id.clone()), new_root_node_rev).await?;

	Ok(AddMapResult { id: map.id.to_string(), root_node_id })
}
